pkg/controller/master/nodedrain/nodedrain_controller.go

```diff
@@ -0,0 +1,293 @@
+	virtualMachineClient         kubevirtv1.VirtualMachineClient
+	virtualMachineCache          kubevirtv1.VirtualMachineCache
+	longhornVolumeCache          longhornv1beta1.VolumeCache
+	longhornReplicaCache         longhornv1beta1.ReplicaCache
+	restConfig                   *rest.Config
+	context                      context.Context
+}
+
+func Register(ctx context.Context, management *config.Management, options config.Options) error {
+	nodes := management.CoreFactory.Core().V1().Node()
+	vmi := management.VirtFactory.Kubevirt().V1().VirtualMachineInstance()
+	vmc := management.VirtFactory.Kubevirt().V1().VirtualMachine()
+	lhv := management.LonghornFactory.Longhorn().V1beta1().Volume()
+	lhr := management.LonghornFactory.Longhorn().V1beta1().Replica()
+	ndc := &ControllerHandler{
+		nodes:                        nodes,
+		nodeCache:                    nodes.Cache(),
+		virtualMachineInstanceCache:  vmi.Cache(),
+		virtualMachineInstanceClient: vmi,
+		virtualMachineClient:         vmc,
+		virtualMachineCache:          vmc.Cache(),
+		longhornReplicaCache:         lhr.Cache(),
+		longhornVolumeCache:          lhv.Cache(),
+		restConfig:                   management.RestConfig,
+		context:                      ctx,
+	}
+
+	nodes.OnChange(ctx, nodeDrainController, ndc.OnNodeChange)
+	return nil
+}
+
+// OnNodeChange handles reconcile logic for node drains
+func (ndc *ControllerHandler) OnNodeChange(key string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.DeletionTimestamp != nil {
+		return node, nil
+	}
+	_, ok := node.Annotations[drainhelper.DrainAnnotation]
+	_, forced := node.Annotations[drainhelper.ForcedDrain]
+
+	if ok {
+		meetsRequirement, err := ndc.meetsControlPlaneRequirements(node)
+		if err != nil {
+			return node, err
+		}
+
+		if !meetsRequirement {
+			return node, fmt.Errorf("unable to place node %s in maintenance mode as this is the only controlplane node, or another controlplane node is already in maintenance mode", node.Name)
+		}
+
+		logrus.Infof("attempting to place node %s in maintenance mode", node.Name)
+		if forced {
+			shutdownVMs, err := ndc.listVMI(node)
+			if err != nil {
+				return node, fmt.Errorf("error listing VMIs in scope for shutdown: %v", err)
+			}
+
+			for _, v := range shutdownVMs {
+				// fetch VMI again in case its changed
+				err := ndc.findAndStopVM(v)
+				if err != nil {
+					return node, err
+				}
+			}
+		}
+		// run node drain
+		err = drainhelper.DrainNode(ndc.context, ndc.restConfig, node)
+		if err != nil {
+			return node, err
+		}
+
+		// remove drain annotation and apply maintenance mode annotations
+		nodeObj, err := ndc.nodeCache.Get(node.Name)
+		if err != nil {
+			return node, err
+		}
+
+		nodeObj.Annotations[ctlnode.MaintainStatusAnnotationKey] = ctlnode.MaintainStatusRunning
+		delete(nodeObj.Annotations, drainhelper.DrainAnnotation)
+		if forced {
+			delete(nodeObj.Annotations, drainhelper.ForcedDrain)
+		}
```

Comment:
We can just delete `ForcedDrain` here. Don't need to check `forced`.

Ref: https://go.dev/play/p/gpKhem3JWAQ

Reply:
👍